File: tests/gpu/consulate_test.py
"""THE CONSULATE'S EMPIRE-WIDE HALF — the GPU side.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/consulate_test.py

The TS twin is tests/cpu/units/consulate.test.ts.

CIV6 (Consulate): "+2 Influence Points per turn. Enemy Spy's level is reduced
by 1 when targeting this city OR CITIES WITH ENCAMPMENTS." The second half is
EMPIRE-WIDE: the building stands in one city and covers every other city of
the same seat that holds a live Encampment.

No scripted lane reaches it — it wants a Diplomatic Quarter with a Consulate in
one city and a finished Encampment in another, of the same seat, with an enemy
spy working the second.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import plant_city, settle_all


def build(rules, path) -> BatchSim:
    return settle_all(BatchSim([load_fixture(path)], rules, device="cpu", dtype=torch.float64))


def two_city_row(sim) -> tuple[int, int, int]:
    """A seat row holding two living cities — the second planted over the
    engine's own FOUND verb, because a bare `step()` founds nothing."""
    row = next(r for r in range(sim.n_majors) if bool(sim.city_alive[0, r, 0]))
    if not bool(sim.city_alive[0, row, 1]):
        plant_city(sim, row)
    assert bool(sim.city_alive[0, row, 1]), f"seat {row} could not be given a second city"
    return row, 0, 1


def put_district(sim, row: int, j: int, di: int) -> int:
    """Give city (row, j) a COMPLETE district of catalog index `di`, on a plot
    the city itself works — `_dist_counts` reads the TILE plane through
    `city_slot_at`, so the plot has to answer with this city's column."""
    sl = sim.city_slot_at(row)
    free = [t for t in range(sim.T)
            if int(sl[0, t]) == j and int(sim.district[0, t]) < 0
            and int(sim.built_wonder[0, t]) < 0 and bool(sim.passable[0, t])
            and t != int(sim.city_center[0, row, j])]
    assert free, f"city ({row}, {j}) works no free plot"
    t = free[0]
    sim.district[0, t] = di
    sim.district_complete[0, t] = True
    sim.district_pillaged[0, t] = False
    sim.city_dist_tile[0, row, j, di] = t
    sim._eff_version += 1
    sim._tile_owner_ver += 1
    return t


def didx(sim, name: str) -> int:
    return [d["id"] for d in sim.districts_cat].index(name)


BLDG_IDS = [b["id"] for b in json.loads(
    (Path(__file__).resolve().parent.parent.parent / "seeder" / "worlds"
     / "rules.json").read_text())["buildings"]]


def bidx(name: str) -> int:
    return BLDG_IDS.index(name)


# ---------------------------------------------------------------------------


def test_wire(rules, path) -> None:
    sim = build(rules, path)
    con = bidx("CONSULATE")
    wide = sim._b_spy_pen_enc.nonzero(as_tuple=True)[0].tolist()
    assert wide == [con], (
        f"the empire-wide half is carried by {[BLDG_IDS[i] for i in wide]}, "
        "and only the Consulate should carry it")
    assert int(sim._b_spy_pen_enc[con]) == 1
    assert int(sim._b_spy_pen[con]) == 1, "the Consulate lost its own-city half"
    print("  1 wire OK — the Consulate alone carries the empire-wide column")


def test_it_reaches_the_other_city(rules, path) -> None:
    sim = build(rules, path)
    row, a, b = two_city_row(sim)
    dq, enc = didx(sim, "DIPLOMATIC_QUARTER"), didx(sim, "ENCAMPMENT")
    con = bidx("CONSULATE")

    base_a = sim._counter_levels(0, row, a)
    base_b = sim._counter_levels(0, row, b)

    put_district(sim, row, a, dq)
    sim.city_bldg[0, row, a, con] = True
    # the Consulate's own city: the Quarter's 2 and the building's 1
    own = sim._counter_levels(0, row, a)
    assert own == base_a + 3, f"the Consulate's own city reads {own}, not {base_a + 3}"
    assert sim._counter_levels(0, row, b) == base_b, "it reached a city with no Encampment"

    put_district(sim, row, b, enc)
    got = sim._counter_levels(0, row, b)
    want = base_b + int(sim._d_spy_pen[enc]) + 1
    assert got == want, f"the Encampment city reads {got}, not {want}"
    # ...and the building's own city still counts it exactly once
    assert sim._counter_levels(0, row, a) == own, "the empire-wide half double-counted at home"
    print(f"  2 reach OK — the other city gains 1 ({base_b} -> {got}), home unchanged at {own}")


def test_a_dead_district_pays_nothing(rules, path) -> None:
    sim = build(rules, path)
    row, a, b = two_city_row(sim)
    dq, enc = didx(sim, "DIPLOMATIC_QUARTER"), didx(sim, "ENCAMPMENT")
    con = bidx("CONSULATE")
    base_b = sim._counter_levels(0, row, b)

    dq_t = put_district(sim, row, a, dq)
    sim.city_bldg[0, row, a, con] = True
    enc_t = put_district(sim, row, b, enc)
    live = sim._counter_levels(0, row, b)
    assert live > base_b

    # an UNFINISHED Encampment is not one
    sim.district_complete[0, enc_t] = False
    assert sim._counter_levels(0, row, b) == base_b, "an unfinished Encampment drew the Consulate"
    sim.district_complete[0, enc_t] = True

    # a PILLAGED Encampment is not one
    sim.district_pillaged[0, enc_t] = True
    assert sim._counter_levels(0, row, b) == base_b, "a pillaged Encampment drew the Consulate"
    sim.district_pillaged[0, enc_t] = False

    # ...and a Consulate whose own Quarter is pillaged pays nowhere
    sim.district_pillaged[0, dq_t] = True
    assert sim._counter_levels(0, row, b) == base_b, "a dark Consulate still paid abroad"
    print("  3 dead districts OK — unfinished, pillaged, and a dark Consulate all pay nothing")


def test_a_rival_consulate_pays_nothing(rules, path) -> None:
    sim = build(rules, path)
    row, a, b = two_city_row(sim)
    other = next(r for r in range(sim.n_majors)
                 if r != row and bool(sim.city_alive[0, r].any()))
    oj = int(sim.city_alive[0, other].nonzero(as_tuple=True)[0][0])
    dq, enc = didx(sim, "DIPLOMATIC_QUARTER"), didx(sim, "ENCAMPMENT")
    con = bidx("CONSULATE")

    base_b = sim._counter_levels(0, row, b)
    put_district(sim, other, oj, dq)
    sim.city_bldg[0, other, oj, con] = True
    put_district(sim, row, b, enc)
    got = sim._counter_levels(0, row, b)
    want = base_b + int(sim._d_spy_pen[enc])
    assert got == want, f"a rival empire's Consulate reached this city ({got} vs {want})"
    print("  4 rival OK — the half is the SEAT's, and stops at its own empire")


def main() -> int:
    rules = load_rules()
    path = fixture_paths()[0]
    test_wire(rules, path)
    test_it_reaches_the_other_city(rules, path)
    test_a_dead_district_pays_nothing(rules, path)
    test_a_rival_consulate_pays_nothing(rules, path)
    print("BATTERY OK consulate")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/cpu/units/consulate.test.ts
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, tileAtCoords, settleAt } from '../helpers';
import { emptySeat, setTileOwner, seatOf } from '../../../cpu/core/seats';
import { cityCounterLevels } from '../../../cpu/core/espionage';
import { BUILDINGS } from '../../../cpu/data/buildings';
import type { City, GameState } from '../../../cpu/core/types';

// CIV6 (Consulate): "+2 Influence Points per turn. Enemy Spy's level is
// reduced by 1 when targeting this city OR CITIES WITH ENCAMPMENTS." The
// second half is empire-wide: the building stands in one city and covers
// every OTHER city of the seat that holds a live Encampment.

/** Give a city a COMPLETE district of `type` on a free owned plot. */
function district(state: GameState, city: City, type: string): number {
  const ctr = state.map.tiles[city.centerIndex];
  const t = state.map.tiles.find(
    (x) => x.index !== ctr.index && !x.district && x.terrain !== 'OCEAN'
      && Math.abs(x.col - ctr.col) <= 2 && Math.abs(x.row - ctr.row) <= 2,
  )!;
  setTileOwner(t, city.seat, city.id);
  t.district = type as never;
  t.districtComplete = true;
  city.districts.push({ type: type as never, tileIndex: t.index });
  return t.index;
}

function world() {
  const state = makeState(makeMap(24, 24));
  state.seats.push(emptySeat(1));
  const capital = settleAt(state, tileAtCoords(state.map, 4, 4).index, 1);
  const other = settleAt(state, tileAtCoords(state.map, 12, 12).index, 1);
  return { state, capital, other };
}

describe('the Consulate row', () => {
  it('carries both halves of the published text', () => {
    expect(BUILDINGS.CONSULATE.influencePerTurn).toBe(2);
    expect(BUILDINGS.CONSULATE.spyLevelPenalty).toBe(1);
    expect(BUILDINGS.CONSULATE.spyLevelPenaltyEncampment).toBe(1);
  });

  it('and it is the only building with the empire-wide half', () => {
    const wide = Object.values(BUILDINGS).filter((b) => (b.spyLevelPenaltyEncampment ?? 0) > 0);
    expect(wide.map((b) => b.id)).toEqual(['CONSULATE']);
  });
});

describe('what the Consulate covers', () => {
  it('its own city, with no Encampment anywhere', () => {
    const { state, capital, other } = world();
    district(state, capital, 'DIPLOMATIC_QUARTER');
    capital.buildings.push('CONSULATE');
    // the Diplomatic Quarter itself takes 2 off, the Consulate standing in it 1
    expect(cityCounterLevels(state, capital)).toBe(3);
    expect(cityCounterLevels(state, other)).toBe(0);
  });

  it('...and the OTHER city, once that one holds a live Encampment', () => {
    const { state, capital, other } = world();
    district(state, capital, 'DIPLOMATIC_QUARTER');
    capital.buildings.push('CONSULATE');
    district(state, other, 'ENCAMPMENT');
    expect(cityCounterLevels(state, other)).toBe(1);
    // ...and the Consulate's own city still counts it exactly once
    expect(cityCounterLevels(state, capital)).toBe(3);
  });

  it('...counting once per Consulate, not once per Encampment', () => {
    const { state, capital, other } = world();
    district(state, capital, 'DIPLOMATIC_QUARTER');
    capital.buildings.push('CONSULATE');
    district(state, other, 'ENCAMPMENT');
    // a THIRD city with an Encampment changes nothing for `other`
    const third = settleAt(state, tileAtCoords(state.map, 18, 18).index, 1);
    district(state, third, 'ENCAMPMENT');
    expect(cityCounterLevels(state, other)).toBe(1);
    expect(cityCounterLevels(state, third)).toBe(1);
  });

  it('...and a second Consulate elsewhere stacks', () => {
    const { state, capital, other } = world();
    district(state, capital, 'DIPLOMATIC_QUARTER');
    capital.buildings.push('CONSULATE');
    district(state, other, 'ENCAMPMENT');
    const third = settleAt(state, tileAtCoords(state.map, 18, 18).index, 1);
    district(state, third, 'DIPLOMATIC_QUARTER');
    third.buildings.push('CONSULATE');
    expect(cityCounterLevels(state, other)).toBe(2);
  });
});

describe('what it does not cover', () => {
  it('a city with no Encampment', () => {
    const { state, capital, other } = world();
    district(state, capital, 'DIPLOMATIC_QUARTER');
    capital.buildings.push('CONSULATE');
    district(state, other, 'CAMPUS');
    expect(cityCounterLevels(state, other)).toBe(0);
  });

  it('an Encampment that is not FINISHED', () => {
    const { state, capital, other } = world();
    district(state, capital, 'DIPLOMATIC_QUARTER');
    capital.buildings.push('CONSULATE');
    const t = state.map.tiles[district(state, other, 'ENCAMPMENT')];
    t.districtComplete = false;
    expect(cityCounterLevels(state, other)).toBe(0);
  });

  it('an Encampment that is PILLAGED', () => {
    const { state, capital, other } = world();
    district(state, capital, 'DIPLOMATIC_QUARTER');
    capital.buildings.push('CONSULATE');
    const t = state.map.tiles[district(state, other, 'ENCAMPMENT')];
    t.districtPillaged = true;
    expect(cityCounterLevels(state, other)).toBe(0);
  });

  it('a Consulate whose own Diplomatic Quarter is pillaged', () => {
    const { state, capital, other } = world();
    const dq = state.map.tiles[district(state, capital, 'DIPLOMATIC_QUARTER')];
    capital.buildings.push('CONSULATE');
    district(state, other, 'ENCAMPMENT');
    expect(cityCounterLevels(state, other)).toBe(1);
    dq.districtPillaged = true;
    expect(cityCounterLevels(state, other)).toBe(0);
    expect(cityCounterLevels(state, capital)).toBe(0);
  });

  it('and a RIVAL empire Consulate', () => {
    const { state, other } = world();
    const mine = settleAt(state, tileAtCoords(state.map, 20, 4).index, 0);
    district(state, mine, 'DIPLOMATIC_QUARTER');
    mine.buildings.push('CONSULATE');
    district(state, other, 'ENCAMPMENT');
    expect(seatOf(state, 0)!.cities.length).toBe(1);
    expect(cityCounterLevels(state, other)).toBe(0);
  });
});
